### PR TITLE
nginx: take care not to pull in module sources as runtime deps

### PR DESCRIPTION
**Pull request by waldheinz**
Nginx likes to print the "configured with ..." stuff on startup,
containing the full configure command line. When built with
modules (which it seems to be by default), this causes the
module sources to appear as runtime dependencies. So just use
the remove-references-to script to patch those out.

For a default installation, the rtmp, dav and moreheaders
module sources are gone, for special cases potentially more.

### CODE REVIEW

File: pkgs/servers/http/nginx/generic.nix [hunk 1]
```diff
@@ -158,9 +158,16 @@ stdenv.mkDerivation {
...
-  '';
+  nativeBuildInputs = [ removeReferencesTo ];
+
+  disallowedReferences = map (m: m.src) modules;
+
+  postInstall =
+    let
+      noSourceRefs = lib.concatMapStrings (m: "remove-references-to -t ${m.src} $out/sbin/nginx\n") modules;
+    in noSourceRefs + (if postInstall != null then postInstall else ''
+      mv $out/sbin $out/bin
```
Comment (ajs124) [inline pkgs/servers/http/nginx/generic.nix:169]: why is this needed?

Comment (waldheinz) [reply pkgs/servers/http/nginx/generic.nix:169]: What exactly do you mean? The changed part implements exactly what's described in this PR and the commit message.

Comment (ajs124) [reply pkgs/servers/http/nginx/generic.nix:169]: >       mv $out/sbin $out/bin

Comment (waldheinz) [reply pkgs/servers/http/nginx/generic.nix:169]: Ah, ok, that line. It has only been moved, so it's really not my idea. Don't all derivations provide their executables in 'bin' only? Not sure, sorry.

Comment (ajs124) [reply pkgs/servers/http/nginx/generic.nix:169]: Ah, don't worry. Didn't read the diff well enough.

> Don't all derivations provide their executables in 'bin' only?

They do, but: https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/setup-hooks/move-sbin.sh

This whole thing is kind of weird, tbh. If you feel like it's out of scope, we can go ahead as is, but we could also just change the `postInstall ? null` to `postInstall ? ""` and remove this `mv` and replace the whole `if … then … else` thing with a simple `+`. Although… maybe that needs a `\n` before the `+`? Is that why you're using `concatMapStrings` instead of `concatMapStringsSep`?

Comment (waldheinz) [reply pkgs/servers/http/nginx/generic.nix]: Thanks for the suggestion, I cleaned up the `postInstall` phase accoringly in another commit.

File: pkgs/servers/http/nginx/generic.nix [hunk 2]
```diff
@@ -158,9 +158,14 @@ stdenv.mkDerivation {
...
-  postInstall = if postInstall != null then postInstall else ''
-    mv $out/sbin $out/bin
-  '';
+  nativeBuildInputs = [ removeReferencesTo ];
+
+  disallowedReferences = map (m: m.src) modules;
+
+  postInstall =
+    let
+      noSourceRefs = lib.concatMapStrings (m: "remove-references-to -t ${m.src} $out/sbin/nginx\n") modules;
```
Comment (ajs124) [inline pkgs/servers/http/nginx/generic.nix:167]: You could not `let` this, but just use the `lib.concatMapStrings […]` directly. That's up to your preference though.

One more and hopefully last real question: Is there any chance that any module (we have in nixpkgs) actually depends on this? As in, needs access to its `.src` store-path?

Comment (waldheinz) [reply pkgs/servers/http/nginx/generic.nix:167]: > You could not `let` this, but just use the `lib.concatMapStrings […]` directly. That's up to your preference though.

I like having small, named building blocks. 

> One more and hopefully last real question: Is there any chance that any module (we have in nixpkgs) actually depends on this? As in, needs access to its `.src` store-path?

Sure that's possible. But because of the structure forced on nginx modules (they all have a `config` file in their root folder which fills in some variables and gets interpreted by the main nginx build script, I think it's unlikely. As in "if it works for some modules, it'll probably work for all". I built nginx with a few random modules and the only mention of the of the  `src` store path was indeed in the configure line.

Comment (ajs124) [reply pkgs/servers/http/nginx/generic.nix:167]: We could also try patching it out there then, instead of using `remove-references-to`, which removes it from the the whole binary. Although I'm fine with just going ahead with this and if it does turn out that there is some module that needs a file from its src during runtime, we can still revert and implement that.

Comment (ajs124) [reply pkgs/servers/http/nginx/generic.nix:167]: > I like having small, named building blocks.

And we don't have a style-guide, so neither of us have anything besides "I like this" on their side. You wrote the code, so that's fine by me.
